File: 3600-3699/3614. Process String with Special Operations II/Solution.java
class Solution {
    public char processStr(String s, long k) {
        int n = s.length();
        long[] len = new long[n];
        long cur = 0;

        for (int i = 0; i < n; i++) {
            char c = s.charAt(i);
            if (c == '*') {
                cur = Math.max(0, cur - 1);
            } else if (c == '#') {
                cur = Math.min(cur * 2, (long) 2e15);
            } else if (c != '%') {
                cur++;
            }
            len[i] = cur;
        }

        if (k >= len[n - 1]) return '.';

        for (int i = n - 1; i >= 0; i--) {
            char c = s.charAt(i);

            if (c == '#') {
                long prevLen = (i > 0) ? len[i - 1] : 0;
                if (k >= prevLen) {
                    k -= prevLen;
                }
            } else if (c == '%') {
                long prevLen = (i > 0) ? len[i - 1] : 0;
                k = prevLen - 1 - k;
            } else if (c != '*') {
                if (len[i] - 1 == k) {
                    return c;
                }
            }
        }

        return '.';
    }
}

File: 3600-3699/3614. Process String with Special Operations II/Solution.cpp
class Solution {
public:

    char processStr(string s, long long k) {

        int n = s.size();

        vector<long long> length(n);

        long long currentLength = 0;

        for (int i = 0; i < n; i++) {

            char current = s[i];

            if (current == '*') {

                currentLength = max(0LL, currentLength - 1);
            }
            else if (current == '#') {

                currentLength =
                    min(
                        currentLength * 2,
                        (long long)2e15
                    );
            }
            else if (current != '%') {

                currentLength++;
            }

            length[i] = currentLength;
        }

        if (k >= length[n - 1]) {
            return '.';
        }

        for (int i = n - 1; i >= 0; i--) {

            char current = s[i];

            if (current == '#') {

                long long previousLength =
                    (i > 0 ? length[i - 1] : 0);

                if (k >= previousLength) {
                    k -= previousLength;
                }
            }
            else if (current == '%') {

                long long previousLength =
                    (i > 0 ? length[i - 1] : 0);

                k = previousLength - 1 - k;
            }
            else if (current != '*') {

                if (length[i] - 1 == k) {
                    return current;
                }
            }
        }

        return '.';
    }
};